File: openwrt/luci/luci-app-openfrp/htdocs/luci-static/resources/view/openfrp/cert.js
'use strict';
'require view';
'require rpc';
'require ui';
'require dom';
'require poll';

/*
 * TLS certificates.
 *
 * Issuance is not an rpcd call. It talks to the CA and then waits for DNS
 * propagation, which takes minutes against rpcd's 30 second limit, so it runs
 * as a detached job and this page follows its log.
 */

var callCert = rpc.declare({
	object: 'luci.openfrp',
	method: 'cert',
	params: ['action', 'params', 'payload'],
	expect: {}
});

var callDNS = rpc.declare({
	object: 'luci.openfrp',
	method: 'dns',
	params: ['action', 'params', 'payload'],
	expect: {}
});

var callJobStart = rpc.declare({
	object: 'luci.openfrp', method: 'job_start',
	params: ['kind', 'args'], expect: {}
});

var callJobStatus = rpc.declare({
	object: 'luci.openfrp', method: 'job_status',
	params: ['id', 'offset'], expect: {}
});

function unwrap(res) {
	if (!res || res.error)
		return Promise.reject(new Error((res && res.error) || _('no response')));
	if (res.ok === false)
		return Promise.reject(new Error(res.error || _('the request failed')));
	return res.data;
}

function call(action, params, payload) {
	return callCert(action, params || {}, payload ? JSON.stringify(payload) : '')
		.then(unwrap);
}

function notifyError(err) {
	ui.addNotification(null, E('p', {}, err.message || String(err)), 'error');
}

function button(label, style, onclick) {
	return E('button', { 'class': 'btn ' + (style || ''), 'click': onclick }, label);
}

var state = { orders: [], cas: [], keyTypes: [], accounts: [] };

/* ------------------------------------------------------------------ */

// stateBadge colours an order by what an operator would do about it.
function stateBadge(order) {
	var colour, label;

	switch (order.state) {
		case 'issued':
			// Expiry matters more than the state once issued: a certificate
			// that renewed months ago and is now expiring is still "issued".
			if (order.days_remaining < 0) {
				colour = '#d9534f';
				label = _('Expired');
			} else if (order.days_remaining <= 7) {
				colour = '#f0ad4e';
				label = _('%d days left').format(order.days_remaining);
			} else {
				colour = '#5bb75b';
				label = _('%d days left').format(order.days_remaining);
			}
			break;
		case 'issuing':
			colour = '#428bca'; label = _('Issuing…'); break;
		case 'failed':
			colour = '#d9534f'; label = _('Failed'); break;
		default:
			colour = '#999'; label = _('Not issued yet');
	}

	return E('span', {
		'style': 'padding:2px 8px;border-radius:3px;white-space:nowrap;color:#fff;' +
			'background:' + colour
	}, label);
}

function ordersTable() {
	if (!state.orders.length)
		return [E('p', {}, _('No certificates yet.'))];

	var head = E('tr', { 'class': 'tr table-titles' }, [
		E('th', { 'class': 'th' }, _('Domains')),
		E('th', { 'class': 'th' }, _('Authority')),
		E('th', { 'class': 'th' }, _('Key')),
		E('th', { 'class': 'th' }, _('Renewal')),
		E('th', { 'class': 'th' }, _('State')),
		E('th', { 'class': 'th' }, '')
	]);

	var rows = state.orders.map(function (order) {
		var detail = [E('div', {}, order.domains.join(', '))];
		if (order.last_error)
			detail.push(E('div', {
				'style': 'color:#d9534f;font-size:90%;word-break:break-word'
			}, order.last_error));

		return E('tr', { 'class': 'tr' }, [
			E('td', { 'class': 'td', 'style': 'max-width:22em' }, detail),
			E('td', { 'class': 'td' }, order.ca_label),
			E('td', { 'class': 'td' }, order.key_type),
			E('td', { 'class': 'td' }, order.auto_renew
				? _('Automatic') : E('em', {}, _('Manual'))),
			E('td', { 'class': 'td' }, stateBadge(order)),
			E('td', { 'class': 'td', 'style': 'text-align:right;white-space:nowrap' }, [
				button(order.state === 'issued' ? _('Renew now') : _('Issue'),
					'cbi-button-action', function () { issue(order); }), ' ',
				button(_('History'), '', function () { showEvents(order); }), ' ',
				button(_('Delete'), 'cbi-button-negative', function () {
					deleteOrder(order);
				})
			])
		]);
	});

	return [E('table', { 'class': 'table' }, [head].concat(rows))];
}

/* ------------------------------------------------------------------ */

function orderDialog() {
	function field(label, input, help) {
		var children = [input];
		if (help)
			children.push(E('div', { 'class': 'cbi-value-description' }, help));
		return E('div', { 'class': 'cbi-value' }, [
			E('label', { 'class': 'cbi-value-title' }, label),
			E('div', { 'class': 'cbi-value-field' }, children)
		]);
	}

	var domainsInput = E('textarea', {
		'class': 'cbi-input-textarea', 'rows': 3, 'style': 'width:100%',
		'placeholder': '*.aiqno.com\nacgshop.aiqno.com'
	});
	var emailInput = E('input', {
		'type': 'text', 'class': 'cbi-input-text', 'style': 'width:100%',
		'placeholder': 'ops@example.com'
	});
	var caSelect = E('select', { 'class': 'cbi-input-select' },
		state.cas.map(function (ca) {
			return E('option', { 'value': ca.key }, ca.label);
		}));
	var keySelect = E('select', { 'class': 'cbi-input-select' },
		state.keyTypes.map(function (k) {
			return E('option', { 'value': k.value }, k.label);
		}));
	// "None" means validate over HTTP, which the tunnel server answers on this
	// router's behalf and which needs no credentials for the zone. A wildcard
	// is the one case that cannot use it.
	var accountSelect = E('select', { 'class': 'cbi-input-select' },
		[E('option', { 'value': '' }, _('None — validate over HTTP'))].concat(
			state.accounts.map(function (a) {
				return E('option', { 'value': String(a.id) },
					a.name + ' (' + a.provider_label + ')');
			})));
	var autoRenew = E('input', {
		'type': 'checkbox', 'class': 'cbi-input-checkbox', 'checked': ''
	});

	// External account binding. ZeroSSL and Google Trust Services refuse to
	// issue without it; Let's Encrypt does not use it at all. The rows are
	// hidden rather than absent so switching authority does not relayout.
	var eabKeyInput = E('input', {
		'type': 'text', 'class': 'cbi-input-text', 'style': 'width:100%'
	});
	var eabHmacInput = E('input', {
		'type': 'password', 'class': 'cbi-input-password', 'style': 'width:100%'
	});
	var eabHint = E('div', { 'class': 'cbi-value-description' }, '');

	function selectedCA() {
		return state.cas.filter(function (ca) { return ca.key === caSelect.value; })[0];
	}

	// eabStored is set once the backend confirms credentials are already on
	// file for the chosen authority. They are entered once and reused for
	// every certificate from it, so asking again is both noise and misleading.
	var eabStored = false;

	function refreshEAB() {
		var ca = selectedCA();
		var needed = !!(ca && ca.requires_eab);

		eabRows.forEach(function (row) { row.style.display = needed ? '' : 'none'; });
		if (!needed)
			return;

		var help = ca.key === 'zerossl'
			? 'https://app.zerossl.com/developer'
			: 'https://console.cloud.google.com/security/publicca';

		// Ask before rendering: the answer decides whether these are fields
		// the operator must fill in or ones they can ignore.
		call('eab-status', { ca: ca.key })
			.then(function (status) {
				eabStored = !!(status && status.present);

				eabKeyInput.placeholder = eabStored ? _('Saved — leave blank to keep') : '';
				eabHmacInput.placeholder = eabStored ? _('Saved — leave blank to keep') : '';

				dom.content(eabHint, eabStored
					? [E('em', {}, _('Already saved for %s and reused automatically. ' +
						'Fill these in only to replace them.').format(ca.label))]
					: [
						_('%s needs a key pair from your %s account.')
							.format(ca.label, ca.label), ' ',
						E('a', {
							'href': help, 'target': '_blank', 'rel': 'noreferrer'
						}, _('Where to find it'))
					]);
			})
			.catch(function () {
				eabStored = false;
				dom.content(eabHint, _('%s needs external account binding credentials.')
					.format(ca.label));
			});
	}

	function save() {
		var domains = domainsInput.value.split(/[\s,]+/).filter(function (d) {
			return d.length > 0;
		});

		if (!domains.length) {
			notifyError(new Error(_('List at least one domain.')));
			return;
		}
		if (!accountSelect.value && domains.some(function (d) {
			return d.indexOf('*') === 0;
		})) {
			notifyError(new Error(
				_('A wildcard can only be proved through DNS. Select an account.')));
			return;
		}

		var ca = selectedCA();
		var wantsEAB = !!(ca && ca.requires_eab);
		var givingEAB = wantsEAB && (eabKeyInput.value || eabHmacInput.value);

		if (wantsEAB && !eabStored && !givingEAB) {
			notifyError(new Error(
				_('%s needs external account binding credentials.').format(ca.label)));
			return;
		}
		if (givingEAB && (!eabKeyInput.value || !eabHmacInput.value)) {
			notifyError(new Error(_('Both fields are required.')));
			return;
		}
		if (!emailInput.value) {
			notifyError(new Error(_('Enter a contact email address.')));
			return;
		}

		// Store the binding first, and only when new values were typed in.
		// Doing it after would leave an order that exists and cannot issue if
		// this call failed; doing it always would overwrite a stored pair with
		// the empty boxes the operator never touched.
		var prepared = givingEAB
			? call('eab', {}, {
				ca: caSelect.value, email: emailInput.value,
				key_id: eabKeyInput.value, hmac: eabHmacInput.value
			})
			: Promise.resolve();

		prepared.then(function () {
		call('order-add', {}, {
			domains: domains,
			key_type: keySelect.value,
			ca: caSelect.value,
			email: emailInput.value,
			account_id: parseInt(accountSelect.value, 10) || 0,
			auto_renew: autoRenew.checked
		})
			.then(function () { ui.hideModal(); refresh(); })
			.catch(notifyError);
		}).catch(notifyError);
	}

	var eabRows = [
		field(_('EAB key ID'), eabKeyInput),
		field(_('EAB HMAC key'), eabHmacInput, eabHint)
	];

	caSelect.addEventListener('change', refreshEAB);

	ui.showModal(_('Request a certificate'), [
		field(_('Domains'), domainsInput,
			_('One per line. \"*\" covers one level only: *.aiqno.com, not a.b.aiqno.com.')),
		field(_('Contact email'), emailInput,
			_('The authority requires one, and uses it for expiry warnings.')),
		field(_('Authority'), caSelect),
		field(_('Key type'), keySelect),
		field(_('DNS account'), accountSelect,
			_('How ownership is proved. HTTP needs the name already pointing here; wildcards need DNS.')),
		eabRows[0],
		eabRows[1],
		field(_('Renew automatically'), autoRenew,
			_('Renews once fewer than 30 days remain.')),
		E('div', { 'class': 'right', 'style': 'margin-top:1em' }, [
			button(_('Cancel'), '', ui.hideModal), ' ',
			button(_('Create'), 'cbi-button-positive', save)
		])
	]);

	refreshEAB();
}

// eabDialog collects binding credentials for an order that cannot issue
// without them, so a failed order can be repaired rather than recreated.
function eabDialog(order, status, then) {
	var keyInput = E('input', {
		'type': 'text', 'class': 'cbi-input-text', 'style': 'width:100%'
	});
	var hmacInput = E('input', {
		'type': 'password', 'class': 'cbi-input-password', 'style': 'width:100%'
	});

	function row(label, input) {
		return E('div', { 'class': 'cbi-value' }, [
			E('label', { 'class': 'cbi-value-title' }, label),
			E('div', { 'class': 'cbi-value-field' }, input)
		]);
	}

	ui.showModal(_('%s needs an account binding').format(status.ca_label), [
		E('p', {}, _('Stored once and reused for every certificate from this authority.')
			.format(status.ca_label, status.ca_label)),
		status.how_to_get ? E('p', {}, E('a', {
			'href': status.how_to_get, 'target': '_blank', 'rel': 'noreferrer'
		}, _('Where to find it'))) : E('span', {}),
		row(_('EAB key ID'), keyInput),
		row(_('EAB HMAC key'), hmacInput),
		E('div', { 'class': 'right', 'style': 'margin-top:1em' }, [
			button(_('Cancel'), '', ui.hideModal), ' ',
			button(_('Save and issue'), 'cbi-button-positive', function () {
				if (!keyInput.value || !hmacInput.value) {
					notifyError(new Error(_('Both fields are required.')));
					return;
				}
				call('eab', {}, {
					ca: status.ca, email: status.email,
					key_id: keyInput.value, hmac: hmacInput.value
				}).then(function () { ui.hideModal(); then(); })
					.catch(notifyError);
			})
		])
	]);
}

/* ------------------------------------------------------------------ */

// issue checks the order can actually proceed, then runs the job.
//
// Starting a job that is certain to be refused wastes minutes and teaches the
// operator nothing they can act on — which is exactly what happened before
// there was anywhere to enter these credentials.
function issue(order) {
	call('eab-status', { id: String(order.id) })
		.then(function (status) {
			if (status && status.required && !status.present) {
				eabDialog(order, status, function () { startIssue(order); });
				return;
			}
			startIssue(order);
		})
		.catch(function () { startIssue(order); });
}

// startIssue runs the job and follows its log.
function startIssue(order) {
	var offset = 0;
	var output = E('pre', {
		'style': 'max-height:24em;overflow:auto;white-space:pre-wrap;font-size:90%;' +
			'background:#1e1e1e;color:#ddd;padding:0.6em;border-radius:3px'
	}, _('Starting…'));
	var statusLine = E('p', {}, _('Takes a few minutes while the DNS record propagates.'));

	var closeButton = button(_('Close'), '', function () {
		poll.remove(tick);
		ui.hideModal();
		refresh();
	});

	function tick() {
		return callJobStatus(jobId, offset).then(function (res) {
			if (!res || res.error)
				return;

			if (res.log && res.log.length) {
				if (offset === 0)
					dom.content(output, res.log);
				else
					output.appendChild(document.createTextNode(res.log));
				output.scrollTop = output.scrollHeight;
			}
			offset = res.offset || offset;

			if (res.state === 'running')
				return;

			poll.remove(tick);
			statusLine.textContent = res.state === 'succeeded'
				? _('The certificate was issued.')
				: _('Issuance failed — see the output above.');
		});
	}

	var jobId = null;

	ui.showModal(_('Issuing certificate'), [
		statusLine, output,
		E('div', { 'class': 'right', 'style': 'margin-top:1em' }, closeButton)
	]);

	callJobStart('cert_issue', JSON.stringify({ id: order.id }))
		.then(function (res) {
			if (!res || res.error || !res.id) {
				statusLine.textContent = _('Could not start: %s')
					.format((res && res.error) || _('no response'));
				return;
			}
			jobId = res.id;
			poll.add(tick, 2);
			tick();
		});
}

function showEvents(order) {
	call('events', { id: String(order.id), limit: '50' })
		.then(function (events) {
			var body = (events && events.length)
				? E('table', { 'class': 'table' }, events.map(function (event) {
					return E('tr', { 'class': 'tr' }, [
						E('td', { 'class': 'td', 'style': 'white-space:nowrap' },
							new Date(event.CreatedAt * 1000).toLocaleString()),
						E('td', { 'class': 'td' }, event.Kind),
						E('td', { 'class': 'td', 'style': 'word-break:break-word' },
							event.Detail || '')
					]);
				}))
				: E('p', {}, _('Nothing has happened to this certificate yet.'));

			ui.showModal(order.domains.join(', '), [
				body,
				E('div', { 'class': 'right', 'style': 'margin-top:1em' },
					button(_('Close'), '', ui.hideModal))
			]);
		})
		.catch(notifyError);
}

function deleteOrder(order) {
	ui.showModal(_('Delete certificate'), [
		E('p', {}, _('Delete the certificate for %s?').format(order.domains.join(', '))),
		E('p', {}, _('Anything still serving it will keep the copy it already has ' +
			'until it restarts.')),
		E('div', { 'class': 'right' }, [
			button(_('Cancel'), '', ui.hideModal), ' ',
			button(_('Delete'), 'cbi-button-negative', function () {
				call('order-delete', { id: String(order.id) })
					.then(function () { ui.hideModal(); refresh(); })
					.catch(function (err) { ui.hideModal(); notifyError(err); });
			})
		])
	]);
}

/* ------------------------------------------------------------------ */

var ordersHolder = E('div', {});

function refresh() {
	return call('orders').then(function (orders) {
		state.orders = orders || [];
		dom.content(ordersHolder, ordersTable());
	}).catch(function (err) {
		dom.content(ordersHolder, E('div', { 'class': 'alert-message warning' },
			E('p', {}, err.message)));
	});
}

// stylesheet returns the app's shared presentation, loaded as part of the view
// so it applies while one of these pages is open and nowhere else. Dialogs
// render outside this node, but the link is in the document for as long as the
// page is, so they pick it up too.
function stylesheet() {
	return E('link', {
		'rel': 'stylesheet',
		'href': L.resource('openfrp/openfrp.css')
	});
}

return view.extend({
	load: function () {
		return Promise.all([
			call('orders').catch(function (err) { return err; }),
			call('cas').catch(function () { return []; }),
			call('keytypes').catch(function () { return []; }),
			callDNS('accounts', {}, '').then(unwrap).catch(function () { return []; })
		]);
	},

	render: function (data) {
		if (data[0] instanceof Error) {
			return E('div', { 'class': 'alert-message warning' }, [
				E('p', {}, _('Certificate management is unavailable.')),
				E('p', {}, data[0].message)
			]);
		}

		state.orders = data[0] || [];
		state.cas = data[1] || [];
		state.keyTypes = data[2] || [];
		state.accounts = data[3] || [];

		dom.content(ordersHolder, ordersTable());

		var warning = null;
		if (!state.accounts.length)
			warning = E('div', { 'class': 'alert-message warning' },
				E('p', {}, _('No DNS account configured. Wildcards need one — add it on the DNS page.')));

		return E('div', {}, [
			stylesheet(),
			E('h2', {}, _('Certificates')),
			E('p', {}, _('Pushed to the server and hot-loaded. Used by tunnels that terminate TLS there.'))
		].concat(warning ? [warning] : []).concat([
			E('div', { 'class': 'cbi-section' }, [
				ordersHolder,
				E('div', { 'style': 'margin-top:1em' },
					button(_('Request a certificate'), 'cbi-button-add', orderDialog))
			])
		]));
	},

	handleSaveApply: null,
	handleSave: null,
	handleReset: null
});

File: internal/manage/cert.go
package manage

import (
	"context"
	"fmt"
	"log/slog"
	"time"

	"github.com/zoefix/openfrp/internal/cert"
	"github.com/zoefix/openfrp/internal/dns"
	"github.com/zoefix/openfrp/internal/storage/repo"
)

type OrderView struct {
	ID      int64    `json:"id"`
	Domains []string `json:"domains"`
	KeyType string   `json:"key_type"`
	CA      string   `json:"ca"`
	CALabel string   `json:"ca_label"`
	Email   string   `json:"email"`

	AccountID   int64  `json:"account_id"`
	AccountName string `json:"account_name"`

	State     string `json:"state"`
	LastError string `json:"last_error,omitempty"`
	AutoRenew bool   `json:"auto_renew"`

	IssuedAt  int64 `json:"issued_at,omitempty"`
	ExpiresAt int64 `json:"expires_at,omitempty"`

	DaysRemaining int    `json:"days_remaining"`
	Issuer        string `json:"issuer,omitempty"`
	SerialNumber  string `json:"serial_number,omitempty"`
}

func (s *Service) CAs() []cert.CA { return cert.CAs() }

func (s *Service) KeyTypes() []map[string]string {
	out := make([]map[string]string, 0, 4)
	for _, keyType := range cert.KeyTypes() {
		out = append(out, map[string]string{
			"value": string(keyType),
			"label": keyType.Label(),
		})
	}
	return out
}

func (s *Service) ListOrders(ctx context.Context) ([]OrderView, error) {
	orders, err := s.orders.List(ctx)
	if err != nil {
		return nil, err
	}

	names := map[int64]string{}
	if accounts, err := s.accounts.List(ctx); err == nil {
		for _, account := range accounts {
			names[account.ID] = account.Name
		}
	}

	now := time.Now()
	out := make([]OrderView, 0, len(orders))
	for _, order := range orders {
		out = append(out, s.view(order, names, now))
	}
	return out, nil
}

func (s *Service) view(order repo.Order, names map[int64]string, now time.Time) OrderView {
	view := OrderView{
		ID: order.ID, Domains: order.Domains, KeyType: order.KeyType,
		CA: order.CA, CALabel: order.CA, Email: order.Email,
		AccountID: order.AccountID, AccountName: names[order.AccountID],
		State: order.State, LastError: order.LastError,
		IssuedAt: order.IssuedAt, ExpiresAt: order.ExpiresAt,
	}

	if order.AutoRenew != nil {
		view.AutoRenew = *order.AutoRenew
	}
	if ca, known := cert.LookupCA(order.CA); known {
		view.CALabel = ca.Label
	}

	if order.ExpiresAt > 0 {
		view.DaysRemaining = int(time.Unix(order.ExpiresAt, 0).Sub(now).Hours() / 24)
	}

	if len(order.Certificate) > 0 {
		material := &cert.Certificate{FullchainPEM: order.Certificate}
		if err := material.Populate(); err == nil {
			view.Issuer = material.Issuer
			view.SerialNumber = material.SerialNumber
		}
	}

	return view
}

type OrderInput struct {
	Domains   []string `json:"domains"`
	KeyType   string   `json:"key_type"`
	CA        string   `json:"ca"`
	Email     string   `json:"email"`
	AccountID int64    `json:"account_id"`
	AutoRenew *bool    `json:"auto_renew,omitempty"`
}

func (s *Service) CreateOrder(ctx context.Context, in OrderInput) (OrderView, error) {
	domains := cert.NormaliseDomains(in.Domains)
	if len(domains) == 0 {
		return OrderView{}, fmt.Errorf("manage: the order needs at least one domain")
	}
	if in.Email == "" {
		return OrderView{}, fmt.Errorf("manage: the CA requires a contact email address")
	}

	keyType := cert.KeyType(in.KeyType)
	if in.KeyType == "" {
		keyType = cert.KeyType("ec256")
	}
	if !keyType.Valid() {
		return OrderView{}, fmt.Errorf("manage: unsupported key type %q", in.KeyType)
	}

	ca := in.CA
	if ca == "" {
		ca = "letsencrypt"
	}
	if _, known := cert.LookupCA(ca); !known {
		return OrderView{}, fmt.Errorf("manage: unknown certificate authority %q", ca)
	}

	if cert.NeedsDNSChallenge(domains) && in.AccountID == 0 {
		return OrderView{}, fmt.Errorf(
			"manage: a wildcard certificate needs a DNS account, because only " +
				"the DNS-01 challenge can prove a wildcard")
	}
	if in.AccountID == 0 && s.httpSolver == nil {
		return OrderView{}, fmt.Errorf(
			"manage: without a DNS account this certificate is validated over " +
				"HTTP through a tunnel server, and none is configured")
	}
	if in.AccountID != 0 {
		if _, err := s.accounts.Get(ctx, in.AccountID); err != nil {
			return OrderView{}, err
		}
	}

	created, err := s.orders.Create(ctx, repo.Order{
		Domains: domains, KeyType: string(keyType), CA: ca,
		Email: in.Email, AccountID: in.AccountID, AutoRenew: in.AutoRenew,
	})
	if err != nil {
		return OrderView{}, err
	}

	_ = s.orders.AppendEvent(ctx, created.ID, "requested", "")
	return s.view(created, map[int64]string{}, time.Now()), nil
}

func (s *Service) DeleteOrder(ctx context.Context, id int64) error {
	return s.orders.Delete(ctx, id)
}

func (s *Service) OrderEvents(ctx context.Context, id int64, limit int) ([]repo.Event, error) {
	return s.orders.Events(ctx, id, limit)
}

func (s *Service) Material(ctx context.Context, id int64) ([]byte, error) {
	order, err := s.orders.Get(ctx, id)
	if err != nil {
		return nil, err
	}
	if len(order.Certificate) == 0 {
		return nil, fmt.Errorf("manage: order %d has not been issued", id)
	}
	return order.Certificate, nil
}

func (s *Service) Issue(ctx context.Context, id int64, progress func(string)) error {
	if progress == nil {
		progress = func(string) {}
	}

	order, err := s.orders.Get(ctx, id)
	if err != nil {
		return err
	}

	if err := s.orders.SetState(ctx, id, repo.StateIssuing, ""); err != nil {
		return err
	}

	certificate, err := s.issue(ctx, order, progress)
	if err != nil {

		_ = s.orders.SetState(ctx, id, repo.StateFailed, err.Error())
		_ = s.orders.AppendEvent(ctx, id, "failed", err.Error())
		return err
	}

	if err := s.orders.StoreCertificate(ctx, id,
		certificate.FullchainPEM, certificate.PrivateKeyPEM,
		certificate.IssuedAt.Unix(), certificate.NotAfter.Unix()); err != nil {
		return err
	}

	kind := "issued"
	if order.IssuedAt > 0 {
		kind = "renewed"
	}
	_ = s.orders.AppendEvent(ctx, id, kind,
		fmt.Sprintf("valid until %s", certificate.NotAfter.Format(time.RFC3339)))

	progress(fmt.Sprintf("certificate valid until %s",
		certificate.NotAfter.Format("2006-01-02")))
	return nil
}

func (s *Service) issue(ctx context.Context, order repo.Order,
	progress func(string)) (*cert.Certificate, error) {

	ca, known := cert.LookupCA(order.CA)
	if !known {
		return nil, fmt.Errorf("manage: unknown certificate authority %q", order.CA)
	}

	account, err := s.acmeAccount(ctx, ca.Key, order.Email)
	if err != nil {
		return nil, err
	}
	if ca.RequiresEAB && account.EABKeyID == "" {
		return nil, fmt.Errorf(
			"manage: %s will not issue without external account binding credentials",
			ca.Label)
	}

	request := cert.IssueRequest{
		Domains: order.Domains,

		CA:      ca.Key,
		KeyType: cert.KeyType(order.KeyType),
		Account: account,
	}

	switch {
	case order.AccountID != 0:
		provider, err := s.provider(ctx, order.AccountID)
		if err != nil {
			return nil, err
		}

		zone, err := dns.RegistrableZone(ctx, provider, order.Domains[0])
		if err != nil {
			return nil, fmt.Errorf("manage: find the zone hosting %s: %w",
				order.Domains[0], err)
		}
		progress(fmt.Sprintf("proving ownership through DNS zone %s", zone))

		request.Solver = dns.NewSolver(provider, zone)

	case cert.NeedsDNSChallenge(order.Domains):
		return nil, fmt.Errorf(
			"manage: order %d covers a wildcard, which only DNS can prove. "+
				"Point it at a DNS account", order.ID)

	case s.httpSolver != nil:
		progress("proving ownership over HTTP, answered by the tunnel server")

		note, err := s.httpSolver.checkReachable(ctx, order.Domains)
		if err != nil {
			return nil, err
		}
		progress(note)

		request.HTTPSolver = s.httpSolver

	default:
		return nil, fmt.Errorf(
			"manage: order %d has no DNS account and no server to answer an "+
				"HTTP validation", order.ID)
	}

	progress(fmt.Sprintf("requesting %v from %s", order.Domains, ca.Label))
	issuer := cert.NewIssuer(slog.Default())

	certificate, err := issuer.Issue(ctx, request)
	if err != nil {
		return nil, err
	}

	if err := s.saveACMEAccount(ctx, ca.Key, account); err != nil {
		return nil, err
	}

	return certificate, nil
}

func (s *Service) acmeAccount(ctx context.Context, ca, email string) (*cert.Account, error) {
	accounts := repo.NewACMEAccounts(s.db.DB)

	stored, err := accounts.Find(ctx, ca, email)
	if err != nil {

		fresh := &cert.Account{Email: email}
		if keyID, hmac, err := accounts.FindEAB(ctx, ca); err == nil {
			fresh.EABKeyID, fresh.EABHMAC = keyID, hmac
		}
		return fresh, nil
	}

	if stored.EABKeyID == "" {
		if keyID, hmac, err := accounts.FindEAB(ctx, ca); err == nil {
			stored.EABKeyID, stored.EABHMAC = keyID, hmac
		}
	}

	account := &cert.Account{
		Email:      stored.Email,
		PrivateKey: stored.PrivateKey,
		EABKeyID:   stored.EABKeyID,
		EABHMAC:    stored.EABHMAC,
	}
	if err := account.LoadRegistration(stored.Registration); err != nil {
		return nil, err
	}
	return account, nil
}

func (s *Service) saveACMEAccount(ctx context.Context, ca string, account *cert.Account) error {
	accounts := repo.NewACMEAccounts(s.db.DB)

	registration, err := account.MarshalRegistration()
	if err != nil {
		return err
	}

	_, err = accounts.Save(ctx, repo.ACMEAccount{
		CA: ca, Email: account.Email,
		PrivateKey: account.PrivateKey, Registration: registration,
		EABKeyID: account.EABKeyID, EABHMAC: account.EABHMAC,
	})
	return err
}

type EABState struct {
	Required  bool   `json:"required"`
	Present   bool   `json:"present"`
	CA        string `json:"ca"`
	CALabel   string `json:"ca_label"`
	Email     string `json:"email"`
	HowToGet  string `json:"how_to_get,omitempty"`
	AccountID int64  `json:"-"`
}

func (s *Service) EABStatus(ctx context.Context, orderID int64) (EABState, error) {
	order, err := s.orders.Get(ctx, orderID)
	if err != nil {
		return EABState{}, err
	}
	return s.EABStatusFor(ctx, order.CA, order.Email)
}

func (s *Service) EABStatusFor(ctx context.Context, caKey, email string) (EABState, error) {
	ca, known := cert.LookupCA(caKey)
	if !known {
		return EABState{}, fmt.Errorf("manage: unknown certificate authority %q", caKey)
	}

	state := EABState{
		Required: ca.RequiresEAB,
		CA:       ca.Key,
		CALabel:  ca.Label,
		Email:    email,
	}
	if !ca.RequiresEAB {
		return state, nil
	}

	state.HowToGet = eabSource(ca.Key)

	if keyID, _, err := repo.NewACMEAccounts(s.db.DB).FindEAB(ctx, ca.Key); err == nil && keyID != "" {
		state.Present = true
	}
	return state, nil
}

func eabSource(ca string) string {
	switch ca {
	case "zerossl":
		return "https://app.zerossl.com/developer"
	case "google":
		return "https://console.cloud.google.com/security/publicca"
	default:
		return ""
	}
}

func (s *Service) SetEAB(ctx context.Context, ca, email, keyID, hmac string) error {
	accounts := repo.NewACMEAccounts(s.db.DB)

	existing, err := accounts.Find(ctx, ca, email)
	if err != nil {
		existing = repo.ACMEAccount{CA: ca, Email: email}
	}
	existing.EABKeyID = keyID
	existing.EABHMAC = hmac

	_, err = accounts.Save(ctx, existing)
	return err
}
